### ISSUE TITLE
Metal decode: the kernels already beat llama.cpp; ~26-29% of wall time is host-side, and that is the whole remaining gap

### ISSUE THREAD

**Issue by antonellof** [antonellof/ferrox#149]
Measured on the M2 Pro with `FERROX_METAL_GPU_TIMING=1`, which reports `MTLCommandBuffer` `GPUStartTime`/`GPUEndTime` and is load-immune. **GPU time and wall time come from the same process**, so they describe the same workload.

| model | wall ms/tok | GPU ms/tok | host ms/tok | host % | llama.cpp ms/tok | gap | gap if host were 0 |
|---|---:|---:|---:|---:|---:|---:|---:|
| Llama-3.2-1B Q4_K_M | 6.94 | **4.90** | 2.04 | 29% | 6.77 | 1.03× | **0.72×** |
| Gemma-2-2B Q4_K_M | 17.61 | **12.98** | 4.62 | 26% | 14.74 | 1.19× | **0.88×** |

`gap = ferrox_ms / llama_ms`. The gaps line up with the ledger's 0.99× and 1.23×.

## The finding

**ferrox's Metal kernels already finish faster than llama.cpp's entire token.** 4.90 ms of GPU work against llama.cpp's 6.77 ms total on Llama-3.2-1B; 12.98 against 14.74 on Gemma-2-2B. Every millisecond of the remaining gap is host-side, between command buffers.

That reframes Metal parity: it is not a kernel problem. Removing the host cost would put both models comfortably under 1×, and it would move **every** Metal decode row at once rather than one model at a time.

## Caveat, stated up front

The host was NOT quiet — an editor's extension host was holding ~122% of a core — so the wall numbers are inflated and **26–29% is an upper bound**, not a measurement of the host cost on an idle machine. The GPU column is unaffected. The direction is solid; the magnitude needs a re-run on a quiet box before it is quoted anywhere.

## What it is not

- **Not per-token KV re-upload.** `upload_from_host` looked hot in a sampler run; instrumenting the reset path showed it fires **twice per bench run**, not per token. Hypothesis measured and dropped.
- **Not ferrox logic.** Across 5131 main-thread samples, every ferrox frame is ≤10 samples. The main thread sits in Metal driver encode paths (`dispatchThreadgroups`, `emitComputeProgramVariantAndArguments`, `memoryBarrierWithResources`, `MTLResourceListAddResource`).

## Where to look next

The host cost is command-encoding and barrier overhead, so the lever is **fewer encoded dispatches and fewer barriers per token**, not faster Rust. Two concrete things to quantify first:

1. How many dispatches ferrox encodes per decode token versus llama.cpp for the same graph. A 16-layer dense decode looks like roughly 14 ops per layer.
2. Whether `MemRanges::begin_op` emits a barrier per op or only on a real read-after-write. `memoryBarrierWithResources` appears in the profile at a similar weight to the dispatch calls themselves.

## Reproducing

```
cargo build --release -p ferrox-cli --features metal
FERROX_METAL_GPU_TIMING=1 ./target/release/ferrox bench \
  -m models/Llama-3.2-1B-Instruct-Q4_K_M.gguf -p 0 -n 128 -r 1 \
  --n-gpu-layers 99 --max-load 0
```

`--max-load 0` is used deliberately here because this is a ratio within one process, not a publishable tok/s. Do not lift the wall numbers above into `benchmarks/RESULTS.md`.

**Comment by antonellof** [comment]
## Two more hypotheses measured and dropped

**Barrier rate is not a tracking defect.** `FERROX_METAL_BARRIER_LOG=1` on Llama-3.2-1B decode reports **0.99 barriers per op**, and the module's own doc says 1.00 means fully serialised. That looks damning, and the first suspicion was that `MemRanges` compares whole buffers rather than ranges — which it does: `buf_key` is the buffer pointer, so a module named `mem_ranges` never compares a range.

But that is not costing anything here. The call sites already **group** the independent work: the three `encode_matvec` calls for Q, K and V sit inside one `begin_op`/`end_op` pair against separate `q_buf`/`k_buf`/`v_buf`, and the same holds for the FFN's gate/up. What remains between groups is a genuine dependency chain — `x_buf` → q/k/v → attention → o → residual → norm — so consecutive groups really do conflict, and ~1 barrier per group is correct rather than conservative.

Range tracking would therefore buy nothing on this path. It is still a naming defect worth fixing on its own terms (the name asserts an invariant the code does not implement), but it is not a performance lever and should not be attempted as one.

**The op count is the lever, not the hazard policy.** 65907 `begin_op`s over 384 decoded tokens is **~172 groups per token**, on a 16-layer model — about 10.75 groups per layer, each costing a driver dispatch encode plus a barrier. Reducing the host cost means **fusing ops so there are fewer groups**, not making the tracker smarter.

So the order of work for this issue is:

1. Re-measure the host share on a quiet machine, to replace the 26–29% upper bound with a number.
2. Count groups per layer against llama.cpp's node count for the same graph, to size the ceiling before writing any fusion.
3. Only then fuse.

Step 2 before step 3 — the same rule that caught the "36% GPU utilization" mistake.

**Comment by antonellof** [comment]
## Measured: the host cost is NOT per-dispatch, and I had the accounting backwards

Step 2 of the plan above was "count groups before fusing". Counting was not enough — I derived ~12 µs of host time per group by dividing total host time by group count across two models, which is a correlation, and it is **wrong by a factor of 20**.

The direct experiment: inject N extra dispatches per layer (an `rms_norm` into a dead `x_buf`, verified inert — output byte-identical), and measure the slope. N=32 on Gemma-2-2B adds 832 dispatches per token, which is large enough that this machine's noise cannot hide it.

| | GPU ms/tok | wall ms/tok | host ms/tok |
|---|---:|---:|---:|
| N=0 | 12.02 | 16.63 | 4.60 |
| N=32 | 16.32 | 21.43 | 5.11 |

832 extra dispatches cost:

| | total | per dispatch |
|---|---:|---:|
| **GPU** | 4.30 ms | **5.16 µs** |
| **host** | 0.50 ms | **0.61 µs** |

## What this changes

**Fusion is not the lever for host overhead.** At 0.61 µs each, *all* ~515 of Gemma-2's real dispatches account for **0.31 ms of the 4.60 ms** host time — about 7%. The other 93% is fixed per-token cost: command-buffer creation, `commit`, the `waitUntilCompleted` round trip, completion handling. Fusing every op in the model would not touch it.

So the host lever is **overlapping host encode with GPU execution** — encoding token N+1's command buffer while N is still running, or otherwise removing the serialised submit-and-wait — not reducing op count.

**Dispatch overhead is a GPU-side cost, and it is large.** 5.16 µs each × ~515 dispatches ≈ **2.66 ms of the 12.02 ms GPU time, roughly 22%**. That is where fusion pays, and it pays against the GPU column, not the host one. (Caveat: 5.16 µs is a small `rms_norm` dispatch *including* its work, so the pure launch overhead a fusion recovers is somewhat less.)

This also explains why the fusion in b119945 measured nothing: 26 dispatches × 5.16 µs = 0.13 ms of GPU time, about 1.1%, sitting right on the noise floor.

## Revised order of work

1. ~~Count groups~~ — done, and the count was the wrong model.
2. **Host, ~4.3 ms/token fixed:** pipeline the encode against execution. Biggest single item, and independent of the graph.
3. **GPU, ~2.7 ms/token in dispatch overhead:** fuse, targeting the GPU column. Worth roughly 0.13 ms per dispatch removed per layer on a 26-layer model.

Also worth recording for anyone benchmarking here: `suggestd` was measured at **87.6%** of a core during this session, which is why a quiet host is not reliably obtainable on this machine and why the experiment was designed around a large effect rather than a quiet baseline.

**Comment by antonellof** [comment]
Status after #202: **the two remaining levers cannot reach the target, and this issue's host figure was an accounting error.**

The "host = wall minus GPU" split used the GPU time of ONE command buffer per token. A sampled decode token has two: the dense stack, and the lm_head, which had no timing tag. Its GPU time, 1.17 ms/tok on Llama-3.2-1B and 2.78 ms on Gemma-2-2B, was being booked as host time. That is most of the 26 to 29% this issue attributed to the host.

Measured now with both buffers tagged, per token:

| | Llama-3.2-1B | Gemma-2-2B |
|---|---|---|
| stack GPU | ~5.0 ms | ~12.2 ms |
| lm_head GPU | ~1.2 ms | ~2.8 ms |
| encode, both buffers | 0.15 to 0.24 ms | 0.42 to 0.45 ms |
| submit latency | ~0.44 ms | ~0.85 ms |

Argument binding lives inside the encode phase, so packing is bounded by 2 to 3% of wall; encode/execute pipelining is bounded by the same number. Three independent measurements agree on the encode cost. Neither lever was built, on purpose.

What is left, bounded: ~0.2 ms submit latency per command buffer, two per sampled token, so folding the lm_head into the stack buffer is worth at most ~4% on the 1B. Gemma's CPU softcap over 256k logits is 0.65 ms, more than its whole encode phase. And Gemma-2-2B's GPU time alone (14.8 ms) already exceeds llama.cpp's whole token (14.74 ms): **the worst Metal row is a kernel gap, not a host one.** `benchmarks/HISTORY.md` is corrected.

**Comment by antonellof** [comment]
Closed by #208 and the 2026-09-11 re-measurement in #209.

The path here is worth recording because the issue's premise inverted twice. It opened attributing 26 to 29% of wall time to host-side work between command buffers. #156 removed 13% of dispatches and 9% of barriers for 2.3% of host time, retiring dispatch count as the lever. #202 then found the host figure itself was an accounting error: the lm_head runs in a second, untimed command buffer, and its GPU time had been booked as host. Encoding was 2 to 3% of wall by three independent measurements, so argument packing and pipelining were retired unbuilt. Gemma-2-2B's GPU time alone exceeded llama.cpp's whole token: a kernel gap.

#208 attributed it per kernel against Llama-3B as the control. The matrix multiplies, 80% of the stack, were already at parity. RoPE, the norms and d128/d256 attention were 3x to 18x llama.cpp's per-dispatch cost, all serial memory-latency chains, and the host-side logit softcap over 256k logits ran on the CPU. Fixed: stack GPU time -15.5% on Gemma-2-2B, -9.6% on Llama-3B, attention rewrite bit-identical, three template kernels replacing twelve copies.

Measured on a quiet host on 2026-09-11: **Gemma-2-2B decode 1.11x to 0.94x**, Llama-3.2-3B 1.03x to 0.94x, and every one of the 16 comparable Metal decode rows is now faster than llama.cpp, summary 0.60x to 0.96x. Prefill 0.99x to 1.09x.

Left, and not blocking: ~0.2 ms submit latency per command buffer, two per sampled token, worth at most ~4% on a 1B; llama.cpp's split-KV attention at long contexts.
